README.md
# HashRing

The HashRing module provides consistent hashing that is compatible with the
original libketama library that was developed at last.fm. In addition to beeing
compatible with `libketama` it's also compatible with the `hash_ring` module for
Python. See the compatiblity section of the API for more details on this.

### Build status

[![BuildStatus](https://secure.travis-ci.org/3rd-Eden/node-hashring.png?branch=master)](http://travis-ci.org/3rd-Eden/node-hashring)

## Installation

The advised installation of module is done through the Node package manager (npm).

```
npm install hashring --save
```

The `--save` parameter tells npm that it should automatically add the module to
the `dependencies` field in your package.json.

## Usage

```js
var HashRing = require('hashring');
```

The HashRing constructor is designed to handle different argument types as a
consistent hash ring can be use for different use cases. You can supply the
constructor with:

#### String

A single server, possible, but pointless in most cases if you only use one
server, then done use the HashRing at all, it only adds overhead.

```js
var ring = new HashRing('127.0.0.1:11211');
```

#### Array

Multiple servers for the HashRing.

```js
var ring = new HashRing(['127.0.0.1:11211', '127.0.0.2:11211']);
```

#### Object

An Object where the keys of the Object are the servers and the value can be a
`Number` and it will be seen as weight for server. The value can also be an
Object. Where the key can be a weight or a vnode.

Weights or vnodes are used to give servers a bigger distribution in the
HashRing. For example you have 3 servers where you want to distribute your keys
over but not all servers are equal in capacity as 2 of those machines have 200mb
of memory and the other has 3.2 gig of memory. The last server is substantially
bigger and there for should receive a greater distrubtion in the ring.

For a rule of thumb use the amount of memory as weight:

```js
var HashRing = require('hashring');

var ring = new HashRing({
  '127.0.0.1:11211': 200,
  '127.0.0.2:11211': { weight: 200 }, // same as above
  '127.0.0.3:11211': 3200
});
```

If you want create a server with multiple vnodes (virtual nodes):

```js
var HashRing = require('hashring');

var ring = new HashRing({
  '127.0.0.1:11211': { vnodes: 50 },
  '127.0.0.2:11211': { vnodes: 200 },
  '127.0.0.3:11211': { vnodes: 100 }
});
```

### Algorithm

With the second argument you can configure the algorithm that is used to hash
the keys. It defaults to `md5` and can only contain values that are accepted in
Node's `crypto` API. Alternatively you can supply it with a function for a
custom hasher. But do note that the hashValue will be calculated on the result.

### Options

- `vnode count` The amount of virtual nodes per server, defaults to 40 as this
  generates 160 points per server as used by ketama hashing.
- `compatiblity` Allows you to force a compatiblity mode of the HashRing. It
  default to ketama hash rings but if you are coming from a python world you
  might want compatiblity with the `hash_ring` module. There's a small diff
  between `hash_ring` and `ketama` and that's the amount of replica's of a server.
  Ketama uses 4 and `hash_ring` uses 3. Set this to `hash_ring` if you want to
  use 3.
- `replicas` The amount of replicas per server. Defaults to 4.
- `max cache size` We use a simple LRU cache inside the module to speed up
  frequent key lookups, you can customize the amount of keys that need to be
  cached. It defaults to 5000.

```js
'use strict';

// require the module, it returns a HashRing constructor
var HashRing = require('hashring');

// Setup hash rings with your servers, in this example I just assume that all
// servers are equal, and we want to bump the cache size to 10.000 items.
var ring = new HashRing([
    '127.0.0.1',
    '127.0.0.2',
    '127.0.0.3', 
    '127.0.0.4'
  ], 'md5', {
    'max cache size': 10000
  });

// Now we are going to get some a server for a key
ring.get('foo bar banana'); // returns 127.0.0.x

// Or if you might want to do some replication scheme and store/fetch data from
// multiple servers
ring.range('foo bar banana', 2).forEach(function forEach(server) {
  console.log(server); // do stuff with your server
});

// Add or remove a new a server to the ring, they accept the same arguments as
// the constructor
ring.add('127.0.0.7').remove('127.0.0.1');
```

### API's Table of Contents

- [HashRing.continuum()](#hashringcontinuum)
- [HashRing.get()](#hashringgetkey)
- [HashRing.range()](#hashringrangekey-size-unique)
- [HashRing.swap()](#hashringswapfrom-to)
- [HashRing.add()](#hashringaddserver)
- [HashRing.remove()](#hashringremoveserver)
- [HashRing.reset()](#hashringreset)
- [HashRing.end()](#hashringend)

#### HashRing.continuum()

Generates the continuum of server a.k.a the Hash Ring based on their weights and
virtual nodes assigned.

---

#### HashRing.get(**key**)

Find the correct node for the key which is closest to the point after what the
given key hashes to.

- **key** String, Random key that needs to be searched in the hash ring

**returns:** The matching server address.

---

#### HashRing.range(**key**, **size**, **unique**)

Returns a range of servers. Could be useful for replication.

- **key** String, Random key that needs to be searched in the hash ring
- **size** Number, Amount items to be returned (maximum). Defaults to the amount
  of servers that are in the hashring.
- **unique** Boolean, Don't return duplicate servers. Defaults to true.

**returns:** The array of servers that we found.

---

#### HashRing.swap(**from*, **to**)

Hotswap identical servers with each other. This doesn't require the cache to be
completely nuked and the hash ring distribution to be re-calculated.

Please note that removing the server and a new adding server could potentially
create a different distribution.

- **from** String, The server that needs to be replaced
- **to** String. The server that replaces the server

---

#### HashRing.add(**server**)

Add a new server to ring without having to re-initialize the hashring. It
accepts the same arguments as you can use in the constructor.

- **server** Server that need to be added to the ring.

---

#### HashRing.remove(**server**)

Remove a server from the hash ring.

- **server** Server that need to be removed from the ring.

---

#### HashRing.has(**server**)

Checks if a given server exists in the hash ring.

- **server** Server for whose existence we're checking.

---

#### HashRing.reset()

Reset the HashRing and clean up it's references.

---

### HashRing.end()

Reset's the HashRing and closes the ring.

---

#### HashRing.find(**hashValue**) (private)

Finds the correct position of the given hashValue in the hashring.

- **hashValue** Number, The hashValue from the HashRing#hashValue method.

**returns:** Index of the value in the ring.

---

#### HashRing.hash(**key**) (private)

Generates the hash for the key.

- **key** String, Random key that needs to be hashed.

**returns:** The hashed valued.

---

#### HashRing.digest(**key**) (private)

Digest hash so we can make a numeric representation from the hash. So it can be
fed in to our hashValue.

- **key** String, Random key that needs to be hashed.

**returns:** An array of charCodeAt(0) converted chars.

---

#### HashRing.hashValue(**key**) (private)

Get the hashed value of the given key, it does the digesting, hashing yo.

- **key** String, Random key that needs to be hashed.

**returns:** The hash value of the key.

---

#### HashRing.points(**servers**)

Returns the points per server.

- **servers** Optional server that you want to filter for

**returns:** A Object with server -> Array of points mapping

## Upgrading from 0.0.x to 1.x.x

The 0.0.x releases had some seriouse flaws that causes it to be incompatible
with the 1.0.0 release. These flaws are the reason that 1.0.0 got released. they
are not backwards compatible as they change the way that keys are hashed. The
following incompatible changes have been made for the sake of consistency:

- Only accepts hashers that are build in to node (for now). As it can only
  guarentee proper hashing of values.
- The replace function was actually doing swaps of keys, so it's original
  functionality has been renamed to `swap`. The replace API is now removing the
  given server and adds it again. As this causes the servers to be propperly
  re-hashed.
- The module now requires a C++ compiler to be installed on your server as
  hashing the value requires support for 64bit bitshifting and JavaScript as a
  language only supports 32bit bitshifting.
- It adds 4 replica's instead 3 for the servers. This is how libketama
  originally did it, if you want to have 3 replica's in the hash ring you can
  set the compatiblity option to `hash_ring`.
- The API's have be renamed, deprication notices are added in to place and they
  will be removed in the next minor version bump (1.1.0)
- Added human readable configuration options instead of camelcase. This
  increases readablity of the module
- CRC32 was removed as crypto engine because it was to unstable.


CHANGELOG
Aug 21, 2015 (3.2.0)
- Changes to the .travis.yml
- Binary add-on is removed infavor of a pure JS solution, this also fixes build
  issues with IO.JS 3.0

Unknown, 2015 (3.1.0)
- Replica's can no longer be set to 0, it should be 1 at a minimum.

Jan 16, 2015 (3.1.0)
- Bump NaN dependency.
- Removed 0.6 from Travis as it's getting install errors for certain dependencies
  consider support for 0.6 removed.

Sept 22, 2014 (3.0.0)
- libketama removes the default supplied port before hashing keys, we now also do
  this to ensure libketama compatibility.
- Upgraded all dependencies to the latest versions.

May 17, 2014 (2.0.0)
- Upgraded Nan and removed node 0.4 support.

Oct 16, 2013 (1.0.3)
- Use the bindings module for proper DEBUG builds.

Sep 13, 2013 (1.0.2)
- Node.js 0.11 support

Mar 14, 2013 (1.0.1)
- Node.js 0.10 support

Feb 24, 2013 (1.0.0)
- Refactored the whole driver, these changes are incompatible with the 0.0.x
  releases as they have changed the distribution of keys.

Jan 11, 2013
- Added LRU caching instead of a plain object for caching, this prevents memory
  leaks

May 03, 2012
- Updated test suites
- Fixed zero arguments bug

May 05, 2011
- Added the missing `removeServer` method.

April 22, 2011
- Externalized the bisection and applied indutny / donnerjack's patches.
- Added test suite
- Optimized nodes lookup

April 21, 2011

- Fixed some issues related to the cache addition as the cache needs to be
  cleared when new servers are added and replaced.


tests/api.test.js
describe('HashRing', function () {
  "use strict";

  var Hashring = require('../')
    , assume = require('assume');

  it('exposes the library version', function () {
    assume(Hashring.version).matches(/^\d+\.\d+\.\d+$/);
  });

  it('libmemcached compatiblity', function () {
    var ring = new Hashring({
      '10.0.1.1:11211': 600,
      '10.0.1.2:11211': 300,
      '10.0.1.3:11211': 200,
      '10.0.1.4:11211': 350,
      '10.0.1.5:11211': 1000,
      '10.0.1.6:11211': 800,
      '10.0.1.7:11211': 950,
      '10.0.1.8:11211': 100
    });

    // VDEAAAAA hashes to fffcd1b5, after the last continuum point, and lets us
    // test the boundary wraparound.
    assume(ring.find(ring.hashValue('VDEAAAAA'))).equals(0);
  });

  describe('API', function () {
    it('constructs with a string', function () {
      var ring = new Hashring('192.168.0.102:11212');

      assume(ring.servers).has.length(1);
      assume(ring.ring.length).is.above(1);
    });

    it('constructs with a array', function () {
      var ring = new Hashring([
          '192.168.0.102:11212'
        , '192.168.0.103:11212'
        , '192.168.0.104:11212'
      ]);

      assume(ring.servers).has.length(3);
      assume(ring.ring.length).is.above(1);
    });

    it('constructs with a object', function () {
      var ring = new Hashring({
          '192.168.0.102:11212': 2
        , '192.168.0.103:11212': 2
        , '192.168.0.104:11212': 2
      });

      assume(ring.servers).has.length(3);
      assume(ring.ring.length).is.above(1);

      ring.servers.forEach(function (server) {
        assume(server.weight).is.above(1);
      });
    });

    it('constructs with a object with per node vnodes', function () {
      var ring = new Hashring({
          '192.168.0.102:11212': { 'vnodes': 40 }
        , '192.168.0.103:11212': { 'vnodes': 50 }
        , '192.168.0.104:11212': { 'vnodes': 5 }
      });

      assume(ring.servers).is.length(3);
      assume(ring.ring.length).equals((40 + 50 + 5) * ring.replicas);

      ring = new Hashring({
          '192.168.0.102:11212': { 'vnodes': 4 }
        , '192.168.0.103:11212': { 'vnodes': 3 }
        , '192.168.0.104:11212': { 'vnodes': 5 }
      });

      assume(ring.servers).has.length(3);
      assume(ring.ring.length).equals((4 + 3 + 5) * ring.replicas);
    });

    it('constructs with a default vnode value', function () {
      var ring = new Hashring([
          '192.168.0.102:11212'
        , '192.168.0.103:11212'
      ], 'md5', { 'vnode count': 60 });

      assume(ring.ring.length).equals(60 * ring.replicas * ring.servers.length);
    });

    it('constructs with no arguments', function () {
      var ring = new Hashring();

      assume(ring.servers).has.length(0);
      assume(ring.ring).has.length(0);
    });

    it('accepts different algorithms', function () {
      var ring = new Hashring('192.168.0.102:11212', 'sha1');

      assume(ring.servers).has.length(1);
      assume(ring.algorithm).equals('sha1');
      assume(ring.ring.length).is.above(1);
    });

    it('generates the correct amount of points', function () {
      var ring = new Hashring({
          '127.0.0.1:11211': 600
        , '127.0.0.1:11212': 400
      });

      assume(ring.ring.length).equals(160 * 2);
    });

    describe("#add", function () {
      it('adds server after zero-argument constructor', function () {
        var ring = new Hashring();
        ring.add('192.168.0.102:11212');

        assume(ring.servers).has.length(1);
        assume(ring.ring.length).is.above(1);
      });
    });

    describe('#get', function () {
      it('looks up keys', function () {
        var ring = new Hashring([
            '192.168.0.102:11212'
          , '192.168.0.103:11212'
          , '192.168.0.104:11212'
        ]);

        assume(ring.find(ring.hashValue('foo'))).is.above(-1);

        // NOTE we are going to do some flaky testing ;P
        assume(ring.get('foo')).equals('192.168.0.102:11212');
        assume(ring.get('pewpew')).equals('192.168.0.103:11212');

        // we are not gonna verify the results we are just gonna test if we don't
        // fuck something up in the code, so it throws errors or whatever

        // unicode keys, just because people roll like that
        assume(ring.find(ring.hashValue('привет мир, Memcached и nodejs для победы'))).is.above(-1);

        // other odd keys
        assume(ring.find(ring.hashValue(1))).is.above(-1);
        assume(ring.find(ring.hashValue(0))).is.above(-1);
        assume(ring.find(ring.hashValue([]))).is.above(-1);
        assume(ring.find(ring.hashValue({wtf:'lol'}))).is.above(-1);

        // this should work as both objects are converted to [object Object] by
        // the .toString() constructor
        assume(ring.get({wtf:'lol'})).equals(ring.get({wtf:'amazing .toStringing'}));
      });
    });

    describe('#hashValue', function () {
      it('should create the correct long value for a given key', function () {
        var ring = new Hashring({
            '127.0.0.1:11211': 600
          , '127.0.0.1:11212': 400
        });

        assume(ring.hashValue('test')).equals(3446378249);
      });
    });

    describe('#find', function () {
      it('should find the correct long value for a given key', function () {
        var ring = new Hashring({
            '127.0.0.1:11211': 600
          , '127.0.0.1:11212': 400
        });

        assume(ring.ring[ring.find(ring.hashValue('test'))].value).equals(3454255383);
      });
    });

    describe('#swap', function () {
      it('swaps servers', function () {
        var ring = new Hashring([
              '192.168.0.102:11212'
            , '192.168.0.103:11212'
            , '192.168.0.104:11212'
          ])
          , amazon = ring.get('justdied')
          , skynet = '192.168.0.128:11212';

        ring.swap(amazon, skynet);
        assume(ring.cache.get('justdied')).equals(skynet);

        // After a cleared cache, it should still resolve to the same server
        ring.cache.reset();
        assume(ring.get('justdied')).equals(skynet);
      });
    });

    describe('#remove', function () {
      it('removes servers', function () {
        var ring = new Hashring([
            '192.168.0.102:11212'
          , '192.168.0.103:11212'
          , '192.168.0.104:11212'
        ]);

        ring.remove('192.168.0.102:11212');
        ring.ring.forEach(function (node) {
          assume(node.server).does.not.equal('192.168.0.102:11212');
        });
      });

      it('Removes the last server', function () {
        var ring = new Hashring('192.168.0.102:11212');
        ring.remove('192.168.0.102:11212');

        assume(ring.servers).has.length(0);
        assume(ring.ring).has.length(0);
      });
    });

    describe('#has', function () {
      it('has a server', function () {
        var ring = new Hashring([
            '192.168.0.102:11212'
          , '192.168.0.103:11212'
          , '192.168.0.104:11212'
        ]);

        assume(ring.has('192.168.0.102:11212')).is.true();
      });

      it('does not have a server', function () {
        var ring = new Hashring([
            '192.168.0.102:11212'
          , '192.168.0.103:11212'
          , '192.168.0.104:11212'
        ]);

        assume(ring.has('192.168.0.105:11212')).is.false();
      });
    });

    describe('#range', function () {
      it('returns 20 servers', function () {
        var ring = new Hashring([
            '192.168.0.102:11212'
          , '192.168.0.103:11212'
          , '192.168.0.104:11212'
        ]);

        assume(ring.range('foo', 20, false)).is.length(20);
      });

      it('returns 3 servers as we only want unique servers', function () {
        var ring = new Hashring([
            '192.168.0.102:11212'
          , '192.168.0.103:11212'
          , '192.168.0.104:11212'
        ]);

        assume(ring.range('foo', 20, false)).is.length(20);
      });
    });
  });
});


tests/distribution.test.js
describe('Hashring distributions', function () {
  "use strict";

  var Hashring = require('../')
    , assume = require('assume');

  this.timeout(20000);

  it('hashes to the exact same output as hash_ring for python', function () {
    var fixture = require('fs').readFileSync(__dirname +'/fixtures/hash_ring.txt')
                               .toString().split('\n');

    var ring = new Hashring({
        '0.0.0.1' : 1,
        '0.0.0.2' : 2,
        '0.0.0.3' : 3,
        '0.0.0.4' : 4,
        '0.0.0.5' : 5
    }, 'md5', { compatibility: 'hash_ring' });

    for (var i=0; i < 100000; i++){
      assume(i + ' ' + ring.get(i)).equals(fixture[i]);
    }
  });

  it('hashes to the exact same output as ketama for python', function () {
    var fixture = require('fs').readFileSync(__dirname +'/fixtures/ketama.txt')
                               .toString().split('\n');

    var ring = new Hashring({
        '0.0.0.1' : 1,
        '0.0.0.2' : 2,
        '0.0.0.3' : 3,
        '0.0.0.4' : 4,
        '0.0.0.5' : 5
    }, 'md5');

    for (var i=0; i < 100000; i++){
      assume(i + ' ' + ring.get(i)).equals(fixture[i]);
    }
  });

  it('hashes to the exact same output as pylibmc', function () {
    var fixture = require('fs').readFileSync(__dirname +'/fixtures/pylibmc.txt')
                               .toString().split('\n');

    var ring = new Hashring([
        '0.0.0.1:11211',
        '0.0.0.2:11211',
        '0.0.0.3',
        '0.0.0.4:11213',
        '0.0.0.5:11212'
    ], 'md5', { 'default port': 11211 });

    for (var i=0; i < 100000; i++){
      assume(i + ' ' + ring.get(i)).equals(fixture[i]);
    }
  });

  it('has an even distribution', function () {
    var iterations = 100000
      , nodes = {
            '192.168.0.102:11212': 1
          , '192.168.0.103:11212': 1
          , '192.168.0.104:11212': 1
        }
      , ring = new Hashring(nodes);

    function genCode (length) {
      length = length || 10;
      var chars = "QWERTYUIOPASDFGHJKLZXCVBNMqwertyuiopasdfghjklzxcvbnm1234567890"
        , numChars = chars.length
        , ret = ""
        , i = 0;

      for (; i < length; i++) {
          ret += chars[parseInt(Math.random() * numChars, 10)];
      }

      return ret;
    }

    var counts = {}
      , node
      , i
      , len
      , word;

    for (i = 0, len = nodes.length; i < len; i++) {
        node = nodes[i];
        counts[node] = 0;
    }

    for (i = 0, len = iterations; i < len; i++) {
      word = genCode(10);
      node = ring.get(word);
      counts[node] = counts[node] || 0;
      counts[node]++;
    }

    var total = Object.keys(counts).reduce(function reduce (sum, node) {
      return sum += counts[node];
    }, 0.0);

    var delta = 0.05
      , lower = 1.0 / 3 - 0.05
      , upper = 1.0 / 3 + 0.05;

    for (node in counts) {
      assume(counts[node] / total).is.within(lower, upper);
    }
  });
});


index.js
'use strict';

var SimpleCache = require("simple-lru-cache")
  , parse = require('connection-parse')
  , crypto = require('crypto');

function hashValueHash(a, b, c, d) {
  return ((a << 24) | (b << 16) | (c << 8) | d) >>> 0;
}

/**
 * Add a virtual node parser to the connection string parser.
 *
 * @param {Object} data server data
 * @param {Mixed} value optional value
 * @api private
 */
parse.extension('vnodes', function vnode(data, value) {
  if (typeof value === 'object' && !Array.isArray(value) && 'vnodes' in value) {
    data.vnodes = +value.vnodes || 0;
  } else {
    data.vnodes = 0;
  }
});

/**
 * HashRing implements consistent hashing so adding or removing servers of one
 * slot does not significantly change the mapping of the key to slots. The
 * consistent hashing algorithm is based on ketama or libketama.
 *
 * @constructor
 * @param {Mixed} server Servers that need to be added to the ring.
 * @param {Mixed} algorithm Either a Crypto compatible algorithm or custom hasher.
 * @param {Object} options Optional configuration and options for the ring.
 * @api public
 */
function HashRing(servers, algorithm, options) {
  options = options || {};

  //
  // These properties can be configured
  //
  this.vnode = 'vnode count' in options ? options['vnode count'] : 40;
  this.algorithm = algorithm || 'md5';

  //
  // if the default port is set, and a host uses it, then it is excluded from
  // the hash.
  //
  this.defaultport = options['default port'] || null;

  //
  // There's a slight difference between libketama and python's hash_ring
  // module, libketama creates 160 points per server:
  //
  //   40 hashes (vnodes) and 4 replicas per hash = 160 points per server.
  //
  // The hash_ring module only uses 120 points per server:
  //
  //   40 hashes (vnodes) and 3 replicas per hash = 120 points per server.
  //
  // And that's the only difference between the original ketama hash and the
  // hash_ring package. Small, but important.
  //
  this.replicas = options.compatibility
    ? (options.compatibility === 'hash_ring' ? 3 : 4)
    : ('replicas' in options ? +options.replicas : 4);

  //
  // Replica's cannot be 0 as it means we have nothing to iterate over when
  // creating the initial hash ring.
  //
  if (this.replicas <= 0) this.replicas = 1;

  // Private properties.
  var connections = parse(servers);

  this.ring = [];
  this.size = 0;
  this.vnodes = connections.vnodes;
  this.servers = connections.servers;

  // Set up a ache as we don't want to preform a hashing operation every single
  // time we lookup a key.
  this.cache = new SimpleCache({
    maxSize: 'max cache size' in  options ? options['max cache size'] : 5000
  });

  // Override the hashing function if people want to use a hashing algorithm
  // that is not supported by Node, for example if you want to MurMur hashing or
  // something else exotic.
  if ('function' === typeof this.algorithm) {
    this.hash = this.algorithm;
  }

  // Generate the continuum of the HashRing.
  this.continuum();
}

/**
 * Generates the continuum of server a.k.a. the Hash Ring based on their weights
 * and virtual nodes assigned.
 *
 * @returns {HashRing}
 * @api private
 */
HashRing.prototype.continuum = function generate() {
  var servers = this.servers
    , self = this
    , index = 0
    , total = 0;

  // No servers, bailout.
  if (!servers.length) return this;

  // Generate the total weight of all the servers.
  total = servers.reduce(function reduce(total, server) {
    return total + server.weight;
  }, 0);

  servers.forEach(function each(server) {
    var percentage = server.weight / total
      , vnodes = self.vnodes[server.string] || self.vnode
      , length = Math.floor(percentage * vnodes * servers.length)
      , key
      , x;

    // If you supply us with a custom vnode size, we will use that instead of
    // our computed distribution.
    if (vnodes !== self.vnode) length = vnodes;

    for (var i = 0; i < length; i++) {
      if (self.defaultport && server.port === self.defaultport) {
        x = self.digest(server.host +'-'+ i);
      } else {
        x = self.digest(server.string +'-'+ i);
      }

      for (var j = 0; j < self.replicas; j++) {
        key = hashValueHash(x[3 + j * 4], x[2 + j * 4], x[1 + j * 4], x[j * 4]);
        self.ring[index] = new Node(key, server.string);
        index++;
      }
    }
  });

  // Sort the keys using the continuum points compare that is used in ketama
  // hashing.
  this.ring = this.ring.sort(function sorted(a, b) {
    if (a.value === b.value) return 0;
    else if (a.value > b.value) return 1;

    return -1;
  });

  this.size = this.ring.length;
  return this;
};

/**
 * Find the correct node for the key which is closest to the point after what
 * the given key hashes to.
 *
 * @param {String} key Key who's server we need to figure out.
 * @returns {String} Server address.
 * @api public
 */
HashRing.prototype.get = function get(key) {
  var cache = this.cache.get(key);
  if (cache) return cache;

  var node = this.ring[this.find(this.hashValue(key))];
  if (!node) return undefined;

  this.cache.set(key, node.server);
  return node.server;
};

/**
 * Returns the position of the hashValue in the hashring.
 *
 * @param {Number} hashValue Find the nearest server close to this hash.
 * @returns {Number} Position of the server in the hash ring.
 * @api public
 */
HashRing.prototype.find = function find(hashValue) {
  var ring = this.ring
    , high = this.size
    , low = 0
    , middle
    , prev
    , mid;

  // Preform a search on the array to find the server with the next biggest
  // point after what the given key hashes to.
  while (true) {
    mid = (low + high) >> 1;

    if (mid === this.size) return 0;

    middle = ring[mid].value;
    prev = mid === 0 ? 0 : ring[mid - 1].value;

    if (hashValue <= middle && hashValue > prev) return mid;

    if (middle < hashValue) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }

    if (low > high) return 0;
  }
};

/**
 * Generates a hash of the string.
 *
 * @param {String} key
 * @returns {String|Buffer} Hash, depends on node version.
 * @api private
 */
HashRing.prototype.hash = function hash(key) {
  return crypto.createHash(this.algorithm).update(key).digest();
};

/**
 * Digest hash so we can make a numeric representation from the hash.
 *
 * @param {String} key The key that needs to be hashed.
 * @returns {Array}
 * @api private
 */
HashRing.prototype.digest = function digest(key) {
  var hash = this.hash(key +'');

  // Support for Node 0.10 which returns buffers so we don't need to charAt
  // lookups.
  if ('string' !== typeof hash) return hash;

  return hash.split('').map(function charCode(char) {
    return char.charCodeAt(0);
  });
};

/**
 * Get the hashed value for the given key.
 *
 * @param {String} key
 * @returns {Number}
 * @api private
 */
HashRing.prototype.hashValue = function hasher(key) {
  var x = this.digest(key);

  return hashValueHash(x[3], x[2], x[1], x[0]);
};

/**
 * None ketama:
 *
 * The following changes are not ported from the ketama algorithm and are hash
 * ring specific. Add, remove or replace servers with as less disruption as
 * possible.
 */

/**
 * Get a range of different servers.
 *
 * @param {String} key
 * @param {Number} size Amount of servers it should return.
 * @param {Boolean} unique Return only unique keys.
 * @return {Array}
 * @api public
 */
HashRing.prototype.range = function range(key, size, unique) {
  if (!this.size) return [];

  size = size || this.servers.length;
  unique = unique || 'undefined' === typeof unique;

  var position = this.find(this.hashValue(key))
    , length = this.ring.length
    , servers = []
    , node;

  // Start searching for servers from the position of the key to the end of
  // HashRing.
  for (var i = position; i < length; i++) {
    node = this.ring[i];

    // Do we need to make sure that we retrieve a unique list of servers?
    if (unique) {
      if (!~servers.indexOf(node.server)) servers.push(node.server);
    } else {
      servers.push(node.server);
    }

    if (servers.length === size) return servers;
  }

  // Not enough results yet, so iterate from the start of the hash ring to the
  // position of the hash ring. So we reach full circle again.
  for (i = 0; i < position; i++) {
    node = this.ring[i];

    // Do we need to make sure that we retrieve a unique list of servers?
    if (unique) {
      if (!~servers.indexOf(node.server)) servers.push(node.server);
    } else {
      servers.push(node.server);
    }

    if (servers.length === size) return servers;
  }

  return servers;
};

/**
 * Returns the points per server.
 *
 * @param {String} server Optional server to filter down.
 * @returns {Object} server -> Array(points).
 * @api public
 */
HashRing.prototype.points = function points(servers) {
  servers = Array.isArray(servers) ? servers : Object.keys(this.vnodes);

  var nodes = Object.create(null)
    , node;

  servers.forEach(function servers(server) {
    nodes[server] = [];
  });

  for (var i = 0; i < this.size; i++) {
    node = this.ring[i];

    if (node.server in nodes) {
      nodes[node.server].push(node.value);
    }
  }

  return nodes;
};

/**
 * Hotswap identical servers with each other. This doesn't require the cache to
 * be completely nuked and the hash ring distribution to be re-calculated.
 *
 * Please note that removing the server and a new adding server could
 * potentially create a different distribution.
 *
 * @param {String} from The server that needs to be replaced.
 * @param {String} to The server that replaces the server.
 * @returns {HashRing}
 * @api public
 */
HashRing.prototype.swap = function swap(from, to) {
  var connection = parse(to).servers.pop()
    , self = this;

  this.ring.forEach(function forEach(node) {
    if (node.server === from) node.server = to;
  });

  this.cache.forEach(function forEach(value, key) {
    if (value === from) self.cache.set(key, to);
  }, this);

  // Update the virtual nodes
  this.vnodes[to] = this.vnodes[from];
  delete this.vnodes[from];

  // Update the servers
  this.servers = this.servers.map(function mapswap(server) {
    if (server.string === from) {
      server.string = to;
      server.host = connection.host;
      server.port = connection.port;
    }

    return server;
  });

  return this;
};

/**
 * Add a new server to ring without having to re-initialize the hashring. It
 * accepts the same arguments as you can use in the constructor.
 *
 * @param {Mixed} servers Servers that need to be added to the ring.
 * @returns {HashRing}
 * @api public
 */
HashRing.prototype.add = function add(servers) {
  var connections = Object.create(null);

  // Add the current servers to the set.
  this.servers.forEach(function forEach(server) {
    connections[server.string] = server;
  });

  parse(servers).servers.forEach(function forEach(server) {
    // Don't add duplicate servers
    if (server.string in connections) return;
    connections[server.string] = server;
  });

  // Now that we generated a complete set of servers, we can update the re-parse
  // the set and correctly added all the servers again.
  connections = parse(connections);
  this.vnodes = connections.vnodes;
  this.servers = connections.servers;

  // Rebuild the hash ring.
  this.reset();
  return this.continuum();
};

/**
 * Remove a server from the hashring.
 *
 * @param {Mixed} server The sever we want to remove.
 * @returns {HashRing}
 * @api public
 */
HashRing.prototype.remove = function remove(server) {
  var connection = parse(server).servers.pop();

  delete this.vnodes[connection.string];
  this.servers = this.servers.map(function map(server) {
    if (server.string === connection.string) return undefined;

    return server;
  }).filter(Boolean);

  // Rebuild the hash ring
  this.reset();
  return this.continuum();
};

/**
 * Checks if a given server exists in the hash ring.
 *
 * @param {String} server Server for whose existence we're checking
 * @returns {Boolean} Indication if we have that server.
 * @api public
 */
HashRing.prototype.has = function add(server) {
  for (var i = 0; i < this.ring.length; i++) {
    if (this.ring[i].server === server) return true;
  }

  return false;
};

/**
 * Reset the HashRing to clean up all references
 *
 * @returns {HashRing}
 * @api public
 */
HashRing.prototype.reset = function reset() {
  this.ring.length = 0;
  this.size = 0;
  this.cache.reset();

  return this;
};

/**
 * End the hashring and clean up all of it's references.
 *
 * @returns {HashRing}
 * @api public
 */
HashRing.prototype.end = function end() {
  this.reset();

  this.vnodes = {};
  this.servers.length = 0;

  return this;
};

/**
 * A single Node in our hash ring.
 *
 * @constructor
 * @param {Number} hashvalue
 * @param {String} server
 * @api private
 */
function Node(hashvalue, server) {
  this.value = hashvalue;
  this.server = server;
}

//
// Set up the legacy API aliases. These will be deprecated in the next release.
//
[
  { from: 'replaceServer' },
  { from: 'replace' },
  { from: 'removeServer', to: 'remove' },
  { from: 'addServer', to: 'add' },
  { from: 'getNode', to: 'get' },
  { from: 'getNodePosition', to: 'find' },
  { from: 'position', to: 'find' }
].forEach(function depricate(api) {
  var notified = false;

  HashRing.prototype[api.from] = function depricating() {
    if (!notified) {
      console.warn();
      console.warn('[depricated] HashRing#'+ api.from +' is removed.');

      // Not every API has a replacement API that should be used
      if (api.to) {
        console.warn('[depricated] use HashRing#'+ api.to +' as replacement.');
      } else {
        console.warn('[depricated] the API has no replacement');
      }

      console.warn();
      notified = true;
    }

    if (api.to) return HashRing.prototype[api.to].apply(this, arguments);
  };
});

/**
 * Expose the current version number.
 *
 * @type {String}
 * @public
 */
HashRing.version = require('./package.json').version;

/**
 * Expose the module.
 *
 * @api public
 */
module.exports = HashRing;
